crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,557 @@
+    },
+    tokio::{sync::broadcast, task::JoinHandle, time::error::Elapsed},
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+    /// The timeout to wait for the result to be ready
+    pub result_ready_timeout: Duration,
+    /// Maximum capacity of the broadcast channel to store the native prices
+    /// results
+    pub broadcast_channel_capacity: usize,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[allow(dead_code)]
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetching: Sync + Send + NativePriceEstimating {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its native price
+    /// estimator result
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, NativePriceEstimateResult>, PriceEstimationError>;
+}
+
+#[async_trait]
+impl NativePriceEstimating for MockNativePriceBatchFetching {
+    async fn estimate_native_price(&self, token: H160) -> NativePriceEstimateResult {
+        let prices = self.fetch_native_prices(&HashSet::from([token])).await?;
+        prices
+            .get(&token)
+            .cloned()
+            .ok_or(PriceEstimationError::NoLiquidity)?
+    }
+}
+
+/// Buffered implementation that implements automatic batching of
+/// native prices requests.
+#[allow(dead_code)]
+#[derive(Clone)]
+pub struct BufferedRequest<Inner> {
+    config: Configuration,
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+/// Object to map the token with its native price estimator result
+#[allow(dead_code)]
+#[derive(Clone)]
+struct NativePriceResult {
+    token: H160,
+    result: Result<f64, PriceEstimationError>,
+}
+
+#[async_trait]
+impl<Inner> NativePriceEstimating for BufferedRequest<Inner>
+where
+    Inner: NativePriceBatchFetching + Send + Sync + NativePriceEstimating + 'static,
+{
+    /// Request to get estimate prices in a batch
+    async fn estimate_native_price(&self, token: H160) -> NativePriceEstimateResult {
+        // Sends the token for requesting price
+        self.calls.unbounded_send(token).map_err(|e| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "failed to append a new token to the queue: {e:?}"
+            ))
+        })?;
+
+        let mut rx = self.broadcast_sender.subscribe();
+
+        tokio::time::timeout(self.config.result_ready_timeout, async {
```

Comment:
@fleupold  Good question, because I wanted to code it as defensively as possible. So there are 2 timeouts:

- This timeout you are pointing out: It is the time we wait until we receive the **correct** token response
- `revc().await` timeout, it is the timeout we wait to receive **any** response. This is the "unnecessary" timeout, but I wanted to add it, in case that the emitter for whatever reason is not emitting results, so we are not waiting there forever.